junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ScriptExecutionCondition.java

```diff
@@ -62,115 +64,116 @@ public ConditionEvaluationResult evaluateExecutionCondition(ExtensionContext con
-			return Optional.empty();
+			return null;
 		}
 		EnabledIf annotation = enabled.get();
 		String source = createSource(annotation.value());
-		Script script = new Script(annotation, annotation.engine(), source, annotation.reason());
-		return Optional.of(script);
+		return new Script(annotation, annotation.engine(), source, annotation.reason());
 	}
 
 	private String createSource(String[] lines) {
 		return String.join(System.lineSeparator(), lines);
 	}
 
-	private ScriptExecutionManager getScriptExecutionManager(ExtensionContext context) {
-		return context.getRoot().getStore(NAMESPACE).getOrComputeIfAbsent(ScriptExecutionManager.class);
-	}
-
-	private Bindings createBindings(ExtensionContext context) {
-		ScriptAccessor configurationParameterAccessor = new ScriptAccessor.ConfigurationParameterAccessor(context);
-		Bindings bindings = new SimpleBindings();
-		bindings.put(Script.BIND_JUNIT_TAGS, context.getTags());
-		bindings.put(Script.BIND_JUNIT_UNIQUE_ID, context.getUniqueId());
-		bindings.put(Script.BIND_JUNIT_DISPLAY_NAME, context.getDisplayName());
-		bindings.put(Script.BIND_JUNIT_CONFIGURATION_PARAMETER, configurationParameterAccessor);
-		return bindings;
-	}
-
-	ConditionEvaluationResult evaluate(ScriptExecutionManager manager, Script script, Bindings bindings) {
-		if (script == null) {
-			return null;
-		}
-		try {
-			Object result = manager.evaluate(script, bindings);
-			return computeConditionEvaluationResult(script, result);
+	/**
+	 * Evaluates scripts and returns a conditional evaluation result.
+	 */
+	interface Evaluator {
+
+		ConditionEvaluationResult evaluate(ExtensionContext context, List<Script> scripts);
+
+		/**
+		 * Create evaluator via reflection to hide the `javax.script` dependency.
+		 *
+		 * <p>This method may return a {@link ThrowingEvaluator} instance on JREs that
+		 * don't provide the "javax.script" package at all. It also returns such an
+		 * instance on JREs that are launched with an active module system using
+		 * insufficient module graphs, i.e. the application does not read
+		 * {@code java.scripting} module.
+		 *
+		 * @see Class#forName(String)
+		 */
+		static Evaluator forName(String name) {
+			// Assert that "javax.script.ScriptEngine" is loadable via basic reflection.
+			return forName("javax.script.ScriptEngine", name);
 		}
-		catch (ScriptException e) {
-			throw new ScriptEvaluationException("Script evaluation failed for: " + script.getAnnotationAsString(), e);
-		}
-	}
 
-	ConditionEvaluationResult computeConditionEvaluationResult(Script script, Object result) {
-		// Treat "null" result as an error.
-		if (result == null) {
-			throw new ScriptEvaluationException("Script returned `null`: " + script.getAnnotationAsString());
+		static Evaluator forName(String nameOfScriptEngine, String name) {
+			// Assert that precondition name is loadable via basic reflection.
+			try {
+				Class.forName(nameOfScriptEngine);
+			}
+			catch (Throwable cause) {
```

Comment:
Do we really want to catch `Throwable` here? How about `ClassNotFoundException`?

Reply:
Yes we do. The throwable that raised the issue was an instance of "java.lang.NoClassDefFoundError: javax/script/ScriptException" -- `extends LinkageError extends Error extends Throwable`.

Reply:
I never understood the rationale for having `NoClassDefFoundError` _and_ `ClassNotFoundException`...

Just to be safe, we should call `BlacklistedExceptions.rethrowIfBlacklisted(cause)` to not accidentally swallow an `OutOfMemoryError`.

Reply:
Aye. Good catch!

Reply:
Yeah... `NoClassDefFoundError` and `ClassNotFoundException` are confusing and have horribly ambiguous names. (sigh)

`ClassNotFoundException` would be thrown if we were using reflection to load the `ScriptException` class; whereas, in our case we actually compiled against it without errors. So `NoClassDefFoundError` is thrown to signal that the dependency (in the class that we are loading without reflection) cannot be found at runtime.

In any case, I usually forget the difference as well and have to look it up... like [here](http://www.baeldung.com/java-classnotfoundexception-and-noclassdeffounderror). 😉

Reply:
+1 for using `BlacklistedExceptions` though!